[dcl.constexpr] Modernize example of constexpr-usable functions (#6484)
cplusplus/draft: source/declarations.tex

\begin{example}
\begin{codeblock}
constexpr int square(int x) {           // OK
  return x * x;
}
constexpr std::generator<int> iota(int n = 0) {

>     while (true) co_yield n++;          // error: \tcode{co_yield} cannot be used in a constexpr function

Reviewer: I think this example can be improved.
[expr.const] bans `co_await` and `co_yield`,
but [dcl.constexpr]p3 (the paragraph the example is attached to)
bans a constexpr function that is a coroutine.
So I think it might better to use `co_return` instead,
which isn't banned by [expr.const].

}

> constexpr int stinit(int n) {           // \tcode{stinit} is constexpr-usable, but \tcode{stinit}

Reviewer: The paragraph the example is attached to defines _constexpr-suitable_, so I think that's what you meant.
```suggestion
constexpr int stinit(int n) {           // \tcode{stinit} is constexpr-suitable, but \tcode{stinit}
```

  static int value = n;                 // may not be called in a constant expression
  return value;
}

> constexpr int uninit() {                // \tcode{uninit} is constexpr-usable, but a program which calls

Reviewer: ```suggestion
constexpr int uninit() {                // \tcode{uninit} is constexpr-suitable, but a program which calls
```

> constexpr int uninit() {                // \tcode{uninit} is constexpr-usable, but a program which calls
>   struct { int a; } s;                  // \tcode{uninit} in a constant expression is ill-formed
>   return s.a;

Reviewer 2: ```suggestion
constexpr int uninit() {                // OK, but never produces a core constant expression
  struct { int a; } s;
  return s.a;                           // undefined behavior: uninitialized read of \tcode{s.a}
```

}
struct Base {
  constexpr Base() = default;           // OK
};
struct Derived : virtual Base {
  constexpr Derived() = default;        // error: constructor cannot be constexpr in a

> };                                      // class or struct with virtual base classes

Reviewer 3: "class or struct" -> "class"

("class" covers all of union, class, struct)
